src/java/org/apache/cassandra/index/sai/disk/PostingListRangeIterator.java

```diff
@@ -90,7 +90,13 @@ public PostingListRangeIterator(IndexContext indexContext,
     @Override
     protected void performSkipTo(PrimaryKey nextKey)
     {
-        if (skipToToken != null && skipToToken.compareTo(nextKey) >= 0)
+        // We defer the actual skipping logic until the next call to computeNext() because skipping requires reading
+        // from disk in order to get the row id at or after the nextKey. As such, we use compareComparableBytes, which
+        // implements a comparison that is equivalent to comparing PrimaryKey byteComparable representations.
+        // This is necessary to prevent issues that can occur if skipToToken is already set to a token-only PrimaryKey
+        // and the nextKey is a PrimaryKey with partition key information because the compareTo method would consider
+        // these two keys equal.
+        if (skipToToken != null && skipToToken.compareComparableBytes(nextKey) >= 0)
```

Comment:
This is a small thing but we found a bug in OS relating to the comparison done here. It is safer to do:

`if (skipToToken != null && skipToToken.compareComparableBytes(nextKey) > 0)`

This makes sure that if the equality is between a token only key and a clustering key we could end up skipping to the wrong row.

Reply:
> This is a small thing but we found a bug in OS relating to the comparison done here.

This is the bug we're trying to solve in this PR :). I solved it by adding `compareComparableBytes`, which does not have the same risk of incorrectly classifying two PKs as equal, so it is correct to use `>=`. The solution to keep `compareTo` and replace `>=` with `>` makes sense if we assume that we're always calling `skipTo` with a larger PK, but if that is true, why do we have the `compareTo` guard here in the first place?

Reply:
@mike-tr-adamson - do you have a reference to that bug fix? I don't see it changed in `ds-trunk` or in the upstream `trunk` branch.

Reply:
I think it is this one that Mike is currently working on. https://issues.apache.org/jira/browse/CASSANDRA-19011